TextFieldEffects/TextFieldEffects/HoshiTextField.swift

```diff
@@ -12,178 +12,226 @@ import UIKit
+
+   This property applies a color to the complete placeholder string. The default value for this property is a black color.
+   */
+  @IBInspectable dynamic open var placeholderColor: UIColor = .black {
+    didSet {
+      updatePlaceholder()
     }
-    
-    private func layoutPlaceholderInTextRect() {        
-        let textRect = self.textRect(forBounds: bounds)
-        var originX = textRect.origin.x
-        switch self.textAlignment {
-        case .center:
-            originX += textRect.size.width/2 - placeholderLabel.bounds.width/2
-        case .right:
-            originX += textRect.size.width - placeholderLabel.bounds.width
-        default:
-            break
-        }
-        placeholderLabel.frame = CGRect(x: originX, y: textRect.height/2,
-            width: placeholderLabel.bounds.width, height: placeholderLabel.bounds.height)
-        activePlaceholderPoint = CGPoint(x: placeholderLabel.frame.origin.x, y: placeholderLabel.frame.origin.y - placeholderLabel.frame.size.height - placeholderInsets.y)
+  }
+
+  /**
+   The scale of the placeholder font.
+
+   This property determines the size of the placeholder label relative to the font size of the text field.
+   */
+  var placeholderFontSize: CGFloat = 12  {
+    didSet {
+      updatePlaceholder()
+    }
+  }
+  
+  override open var placeholder: String? {
+    didSet {
+      updatePlaceholder()
+    }
+  }
+
+  override open var bounds: CGRect {
+    didSet {
+      frame = CGRect(x: frame.origin.x,
+                     y: frame.origin.y,
+                     width: bounds.size.width,
+                     height: height)
+    }
+  }
+
+  override open func awakeFromNib() {
+    super.awakeFromNib()
+    initPlaceholderFont()
+  }
+
+  private let borderThickness: (active: CGFloat, inactive: CGFloat) = (active: 2, inactive: 0.5)
+  private let placeholderInsets = CGPoint(x: 0, y: 25)
+  private let textFieldInsets = CGPoint(x: 0, y: 6)
+  private let inactiveBorderLayer = CALayer()
+  private let activeBorderLayer = CALayer()
+  private var activePlaceholderPoint: CGPoint = CGPoint(x: 0, y: -2)
+  private var inactivePlaceholderPoint: CGPoint {
+    get {
+      if let font = font, font.pointSize > 15 {
+        return CGPoint(x: 0, y: 14)
+      } else {
+        return CGPoint(x: 0, y: 20)
+      }
+    }
+  }
 
+  private var placeholderLabelOriginalText: String?
+  private var placeholderFont: UIFont?
+  private var height: CGFloat = 45
+
+  open func showError(message: String) {
+    placeholderLabelOriginalText = placeholderLabel.text
+    placeholderLabel.textColor = borderActiveColor
+    placeholderLabel.text = message
+    activeBorderLayer.frame = self.rectForBorder(self.borderThickness.active)
+    activeBorderLayer.isHidden = false
+    placeholderLabel.sizeToFit()
+  }
+
+  open func hideError() {
+    placeholderLabel.text = placeholderLabelOriginalText != nil ? placeholderLabelOriginalText : placeholderLabel.text
```

Comment:
May you can try something shorter version like 
```placeholderLabel.text = placeholderLabelOriginalText ?? placeholderLabel.text```

Reply:
I think I was getting a warning doing that, but I'll try

Reply:
it shows a compilation error, because `placeholderLabelOriginaText` is a `string` not a boolean.
<img width="1354" alt="screen shot 2017-11-27 at 1 16 47 pm" src="https://user-images.githubusercontent.com/4087387/33276769-419ae2c8-d375-11e7-8be5-c18a43df57dd.png">

Reply:
@Sh4rky You have to use the nil coalescing operator `??`, double question marks.